presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleBreakdownHistogram.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.fixedhistogram;
+
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class FixedDoubleBreakdownHistogram
```

Comment:
Can you add a comment explaining the purpose and usage of this type of histogram?  I'm wondering if we need the nested List<Map<Double, Long>> type, and if we can represent this in a more flat way.  I'm also concerned about the use of Double as the key to the map (what if the double value is off by some very small amount, does this class do the right thing?)

Reply:
I think it's a must for implementing the jacknife correction. Inaccuracies of doubles are a great point. They do not alter the correctness, but can incur a performance hit (due to increased size). The documentation states that this method can be expensive if the cardinality of distinct weights is large. For the special, but common, case of weights = 1, this is not an issue.

Reply:
This will complicate the implementation, but we can do it differently, without the need for these nested structural types.

At one part, we have a hash table, where the key of the hash table is a composite key of the histogram bucket and the weight.  The values are incremented as weights are added, with the initial value of the hash table being 1.  This is stored with 3 arrays, one for the buckets, one for the weights, and one for the values, where the total size of these arrays are the number of distinct bucket + weight combinations.

This by itself isn't sufficient because you also need to support iterating over the buckets, and then iterating over the weights + counts within the buckets.  For that we can maintain a linked list, where we have one array which indexes into the initial position of the bucket, and a parallel array of the same size as the hash table where the element points to the next element of the linked list (to retrieve the value, simply retrieve the value from the hash table at the same index position).

This will complicate the implementation, but I think it's worth it, because especially with large structural types this implementation has the potential to generate a large amount of garbage and put GC pressure on Presto clusters.  It will make it safer and more performant.  Of course, if you can think of a simpler approach using flat datastructures I'm all ears, but this is how I'm thinking about it.